Improve: add autowrap and support for Chinese/other language word wrapping

hhsiao opened the pull request:
Modify text wrapping to use actual text width as calculated by "horizontalAdvance", this allows the wrapping to work for a language where character width is not the same across available characters.

<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Text wrapping now uses QFontMetrics(font).horizontalAdvance to determine where the wrap should happen.  

#### Motivation for adding to Mudlet
Text wrapping was assuming all characters are equal width, thus it completely breaks if that isn't the case.

#### Other info (issues closed, discussion etc)
closes https://github.com/Mudlet/Mudlet/issues/5564

hhsiao committed:
Update TBuffer.cpp

Modify text wrapping to use actual text width as calculated by "horizontalAdvance", this allows the wrapping to work for a language where character width is not the same across available characters.

hhsiao commented:
![sample](https://user-images.githubusercontent.com/687505/182439953-6d2df352-e236-4e9a-ab3f-7d714d433315.png)

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2441,75 +2441,113 @@ inline int TBuffer::wrap(int startLine)
+        bool hasContent = false;
+
+        // Track the current width in pixel
+        int lineWidth = QFontMetrics(hostFont).horizontalAdvance(lineBuffer[i]);
```
```suggestion
        int lineWidth = QFontMetrics(hostFont).horizontalAdvance(lineBuffer.at(i));
```

hhsiao replied:
I've made the changes, thanks, solid suggestion :)

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2441,75 +2441,113 @@ inline int TBuffer::wrap(int startLine)
+        int lineWidth = QFontMetrics(hostFont).horizontalAdvance(lineBuffer[i]);
+        if (lineWidth >= wrapByPixel) {
+            // This is where actual wrapping occurs
+            for (int i2 = 0, total = static_cast<int>(buffer[i].size()); i2 < total;) {
```
👎  No. no, no!

This adding a `QChar`/`TChar` at a time to `lineText`/`newLine` is no longer safe to use once we are (and we *are* **now**) handling non-ASCII text. `QString`s contain effectively UTF-16BE encoded text and UTF-16 like UTF-8 is a ***variable length encoding*** system. This means that a single `QChar` may NOT be a single grapheme (what some people think is a "character"), even a pair of them (a high-surrogate followed by a low-surrogate) may not be a grapheme. To step through the "characters" you must step through them "grapheme" by "grapheme" - which means putting the whole line of text (in `lineBuffer.at(i)`) and using `QTextBoundaryFinder` to work out which `QChars` are in each grapheme...

SlySven replied:
You might like to take a look at `(int) TTextEdit::convertMouseXToBufferX(const int mouseX, const int lineNumber, bool* isOutOfbounds, bool* isOverTimeStamp) const` as that does do the right thing as it tries to work out where the mouse is positioned on a line of text on the screen and relate that to the matching horizontal position (`x`) of the start of the grapheme in the `TBuffer::lineBuffer.at(y).at(x)`...

vadi2 replied:
We are very passionate about graphemes here. Thanks @hhsiao for working on this 😬

hhsiao replied:
I've made changes to do wrapping base on QTextBoundary, which I believe would prevent my code from cutting character off at the wrong place

SlySven replied:
😊 Sorry @hhsiao if I came across a bit strong in https://github.com/Mudlet/Mudlet/pull/6207#discussion_r935999160 - TBH this was something that I was planning to work on during my foreign holiday/Summer vacation (starting in a couple of weeks) but you have dived in ahead of me and I wanted to steer you away from considering individual `QChar`s as we have been doing...

I have some thoughts on how we might improve things which I'll raise in a thread in the Discord #mudlet-development channel...

SlySven reviewed (commented):
Sorry, but this needs a rework... 🙁

hhsiao committed:
TBuffer wrapping optimization, generic auto wrapping

TBuffer now uses the more efficient at(i) instead of operator=[i]
Wrapping position calculation now uses QTextBoundaryFinder::word to find boundary
Auto wrapping for wrap setting below 10, and should properly autowrap all mini console.

hhsiao committed:
Merge branch 'development' into Text-Wrapping

hhsiao committed:
Text wrapping for TBuffer::append

Took what I did with TBuffer::wrap and implemented for TBuffer::append

hhsiao committed:
Merge branch 'Text-Wrapping' of https://github.com/hhsiao/Mudlet into Text-Wrapping

hhsiao committed:
Update TBuffer.cpp

Fix CodeFactor issues

hhsiao committed:
Update wrapLine, refactor wrap

wrapLine is a special version of wrap, basically wrap that only wrap one line.
Code in wrap is now refactored into wrapLine.
Also removed two of the debug message, since those were used only for me to figure out what was going on.

hhsiao committed:
Update TBuffer.cpp

Missed a line when I did the refactoring, pSpace should be format that's passed in

hhsiao committed:
Update TBuffer.cpp

Removed unused variables

hhsiao committed:
CodeFactor change

Added space around =

hhsiao committed:
Wrapping bug fixes

Wrapping now accounts for LineFeed, instead of pushing it in, it actually does a new line (which is what line feed is supposed to do)

wrapLine now properly does the single line mode, and uses remove -> insert , as oppose to "pop_back" "push_back".

Change different place where wrapLine was being called with incorrect width.

hhsiao committed:
Update TBuffer.cpp

Added comment explaining what wrapLine does

hhsiao committed:
Update TBuffer.cpp

Change the QTextBoundaryFinder mode from Word to Line

vadi2 committed:
Add UI for autowrap

vadi2 committed:
Re-set tab index to one

SlySven commented on src/Host.h:
```diff
@@ -516,7 +516,11 @@ class Host : public QObject
+    // or wrap at a specific amount of characters:
     int mWrapAt;
+    // after wrapping a line, intent it by how many spaces:
     int mWrapIndentCount;
```
~~This seems to be duplicated 2 lines down - and doesn't match the comment - is this a merge (of `development` branch code into this PR) cock-up?~~

Oh no it isn't - I was only reviewing a sub-set of the changes and wasn't seeing the whole picture. 🤦‍♂️

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2074,19 +2074,37 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
-    bool firstChar = (lineBuffer.back().isEmpty());
+
+    // Whethere this is the first character in the line
+    bool firstChar = (lineBuffer.back().size() == 0);
```
:-1: The previous code (using the `isEmpty()`) is held by us (the Mudlet Core Devs) to be a clearer expression of intent than a `size() == 0` - please could you revert this particular line... :pleading_face:

hhsiao replied:
I've changed all instance of ".size() == 0" to ".isEmpty()" (there was 3)

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2550,209 +2694,82 @@ inline int TBuffer::wrap(int startLine)
-                    for (int i3 = 0; i3 < mWrapIndent; ++i3) {
-                        TChar pSpace;
+                    for (int i3 = 0; i3 < indentSize; ++i3) {
+                        TChar pSpace = format;
```
:+1: Well spotted - I suspect this is why previously, wrapped text (when there was a non-zero indent) was being preceded by black spaces instead of those in the correct background colour... :grinning:

hhsiao replied:
actually everywhere that pass in "format" is actually just passing in a generic pSpace, so this wouldn't fix that.  Now I am interested in fixing that problem, but I believe it isn't related to this PR, so as such I would do a separate PR for that.

vadi2 replied:
Hmm I may not be grasping the problem discussed here, but is it really an issue?

hhsiao replied:
I think what the code "should do", is figure out what the format of the indent "should be".  Now that's what I would need to find out, which is what is a generic space supposed to look like?

SlySven replied:
> actually everywhere that pass in "format" is actually just passing in a generic pSpace, so this wouldn't fix that. Now I am interested in fixing that problem,...

I am not sure that is the case - this method `TBuffer::wrapLine(...)`  is used in more than just `TConsole::luaWrapLine(int)` and in those other cases (`TConsole::printCommand(...)`, `TConsole::insertText(...)`, `TConsole::insertLink(...)`) a format is supplied...

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -128,6 +128,9 @@ TTextEdit::TTextEdit(TConsole* pC, QWidget* pW, TBuffer* pB, Host* pH, bool isLo
 
     connect(mpHost, &Host::signal_changeIsAmbigousWidthGlyphsToBeWide, this, &TTextEdit::slot_changeIsAmbigousWidthGlyphsToBeWide, Qt::UniqueConnection);
+    
+    if (mpHost->mAutoWrap)
```
Minor code-style note - we use `{`...`}` around even a single line of the conditional code for an `if (....)`... - I'd post the code as a suggestion but my Web-browser is playing up and cannot process a multi-line selection currently... :roll_eyes:
```diff
-    if (mpHost->mAutoWrap)
+    if (mpHost->mAutoWrap) {
         mpBuffer->setWrapAt(width() / mFontWidth);
+    }
```

SlySven commented on src/TTextEdit.cpp:
```diff
@@ -290,6 +293,7 @@ void TTextEdit::updateScreenView()
         updateScrollBar(mpBuffer->mCursorY);
```
As I am currently seeing the code it is (from new line 291):
```cpp
    mScreenHeight = visibleRegion().boundingRect().height() / mFontHeight;
    if (!mIsLowerPane) {
        updateScrollBar(mpBuffer->mCursorY);
    }
    int currentScreenWidth = visibleRegion().boundingRect().width() / mFontWidth;

    } else {
        mScreenWidth = currentScreenWidth;
    }
        // Note the values in the "parent" Host instance - for the UPPER pane
        // so that they are available for NAWS:
        if (!mIsLowerPane) {
            mpHost->setScreenDimensions(mScreenWidth, mScreenHeight);
        }
    } else {
        mScreenWidth = currentScreenWidth;
    }

    if (mpHost->mAutoWrap) {
        mpBuffer->setWrapAt(mScreenWidth);
    }

    mOldScrollPos = mpBuffer->getLastLineNumber();
}
```

Either my web browser is really messed up - it could be, GitHub insists on using bleeding edge code and that is not fully standards compliant and breaks on browsers that are not trying to be Chrome-a-likes (i.e. not Chrome/FireFire/Edge?) - or there is something messed up with that chunk...

SlySven replied:
*\*Sigh\** - another case where I wasn't seeing the whole picture. Sorry but there is nothing to see here - just move along please... 🙄

SlySven commented on src/dlgProfilePreferences.cpp:
```diff
@@ -538,6 +540,8 @@ void dlgProfilePreferences::enableHostDetails()
     groupBox_doubleClick->setEnabled(true);
 
+    checkBox_autoWrap->setEnabled(true);
+
```
```suggestion

```

SlySven replied:
Forget the previous change I suggested - again I didn't see the whole thing correctly.

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -42,7 +42,7 @@
       <enum>QTabWidget::Rounded</enum>
      </property>
      <property name="currentIndex">
-      <number>0</number>
```
:information_source: When you save the form/dialogue from the stand-alone Qt Designer utility or the version integrated into the Qt Creator IDE it remembers the current tab you are on - you need to always switch back to the first tab (or avoid this change getting into the end PR) to prevent this (change of the "currentIndex" away from `0`) from appearing...

As it happens I have put code into the `dlgProfilePrefences` constructor to prevent this from having any effect (it forces the current tab to the first one automatically) - but basically this change is not wanted... :grinning:

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -1032,7 +1049,7 @@ you can use it but there could be issues with aligning columns of text</string>
              </item>
              <item>
-              <widget class="QLabel" name="label_23">
+              <widget class="QLabel" name="label_wrapCharacters">
```
:+1: for giving a meaningful name to the label for the control - too many of them have been left at the default of `label_` and a number...! :grinning:

SlySven commented on src/ui/profile_preferences.ui:
```diff
@@ -4145,5 +4162,38 @@ you can use it but there could be issues with aligning columns of text</string>
   <include location="../mudlet.qrc"/>
  </resources>
- <connections/>
+ <connections>
```
:thinking: Whilst it is possible to define a signal-slot connection in the form/dialogue `.ui` file it is not so easy for some of us coders to spot/find them - speaking personally I would prefer it if these could be done via `connect(...)` calls in the `dlgProfilePrefences` constructor. Can I check @vadi2 if you share this view?

vadi2 replied:
I defined them in there but I can move it!

vadi2 committed:
Move connections to code

vadi2 committed:
Add missing brace

vadi2 committed:
Fix autowrap not setting effect properly

vadi2 committed:
Merge branch 'development' into Text-Wrapping

hhsiao committed:
Update src/dlgProfilePreferences.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

hhsiao committed:
Use isEmpty() as oppose to size() == 0

Change instance where we are doing size() == 0, to do isEmpty() instead.

hhsiao commented:
This is ready for review again from my perspective

vadi2 commented:
Thanks, will review. @Mudlet/core-cpp

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2052,15 +2052,15 @@ void TBuffer::resetColors()
         shrinkBuffer();
     }
+
+    // This is index to the previous line
```
```suggestion
    // index to the previous line
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2052,15 +2052,15 @@ void TBuffer::resetColors()
+    // This is index to the previous line
     int last = buffer.size() - 1;
+
+    // Prevent instance where buffer is empty
```
Maybe this?

```suggestion
    // prevent a case where buffer is completely empty
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2074,19 +2074,37 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
         last = 0;
     }
+
+    // Whethere this is the first character in the line
```
```suggestion
    // processing the first character in the line?
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2074,19 +2074,37 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
     bool firstChar = (lineBuffer.back().isEmpty());
+
+    // Length of the text we are appending
     int length = text.size();
```
What do you think of removing the comment and changing the variable name to a self-descriptive `textLength`? 

I can do this change

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2074,19 +2074,37 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
         return;
     }
+
+    // If we are looking to append more character than what is allowed
```
```suggestion
    // limit amount of characters that can be appended at once
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2096,44 +2114,63 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
             continue;
         }
+        
+        // The way wrapping works, is we wait until it's exceeded, and then take away content until it's below the limit
```
```suggestion
        // wait until the text overflows the line, then take content away until it's below the wrap limit
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2096,44 +2114,63 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+        // The way wrapping works, is we wait until it's exceeded, and then take away content until it's below the limit
+        if (lineWidth > wrapByPixel) {
+            QTextBoundaryFinder wordFinder(QTextBoundaryFinder::Line, lineBuffer.back());
+            int lineSize = lineBuffer.back().size();
```
```suggestion
            const int lineSize = lineBuffer.back().size();
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2096,44 +2114,63 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+            QTextBoundaryFinder wordFinder(QTextBoundaryFinder::Line, lineBuffer.back());
+            int lineSize = lineBuffer.back().size();
+            wordFinder.setPosition(lineSize-1);
+            // How character is chopped off (aka wrapped to next line)
```
```suggestion
            // how many characters are chopped off (aka wrapped to next line)
```

?

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2096,44 +2114,63 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+                wrapPos = wordFinder.position();
+            } else {
+                while (!(wordFinder.isAtBoundary() && (wordFinder.boundaryReasons() & QTextBoundaryFinder::BreakOpportunity) != 0)) {
+                    wrapPos =  wordFinder.toPreviousBoundary();
```
```suggestion
                    wrapPos = wordFinder.toPreviousBoundary();
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2096,44 +2114,63 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
+            }
         }
+
+        // Append one character, this is placed here at the very end becuase one character can't be wrapped anyway, so edge case is irrelevant
```
```suggestion
        // Append one character, this is placed here at the very end because one character can't be wrapped anyway, so edge case is irrelevant
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2096,44 +2114,63 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
                 }
             }
+
+            // We chop off however many character we are supposed to, off of lineBuffer
```
```suggestion
            // We chop off however many characters we are supposed to, off of lineBuffer
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2148,15 +2185,17 @@ void TBuffer::append(const QString& text, int sub_start, int sub_end, TChar form
     }
 }
 
+// This function adds text to existing buffer, it needs to ensure wrapping is done properly
```
```suggestion
// adds text to existing buffer and ensures wrapping is done properly
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2611,59 +2587,60 @@ void TBuffer::logRemainingOutput()
+        bool isPrompt = promptBuffer.at(i);
         std::deque<TChar> newLine;
-        QString lineText;
+        QString lineText = "";
```
```suggestion
        QString lineText;
```
Does this also work? QString starts empty by default

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2672,49 +2649,127 @@ int TBuffer::wrapLine(int startLine, int screenWidth, int indentSize, TChar& for
+                    timeList.append(time);
+                    promptList.append(isPrompt);
+                    newLine.clear();
+                    lineText = "";
```
```suggestion
                    lineText = QString();
```

Qt-recommended way of zeroing out a string

vadi2 committed:
Fix crash with existing profiles

hhsiao committed:
TBuffer.cpp Text Wrapping Optimization

QFontMetrics is now only initialized once
QTextBoundaryFinder is only initialized once per line
No longer building wrapped line 1 character at a time, using subString instead

hhsiao committed:
Optimization of text wrapping

Wrapping now uses QStringRef

hhsiao committed:
CodeFactor changes

hhsiao committed:
Further optimization

Moved variables to the innermost scope
Reduced number of variable calculation
Reduced number of variable assignment
Use --X as oppose to X-- (pre-increment vs post-increment)
Change variable name to be more meaningful
Use at(x) as oppose to [x]

hhsiao committed:
Wrapping Optimization - Binary Search for edge case

For edge case where boundary span across the entire line, use binary search as oppose to building the string one character at a time

hhsiao committed:
Append and AppendLine

Apped and AppendLine now use the same wrapLine function.

hhsiao committed:
CodeFactor fix

CodeFactor fix, missing space before ( in while

hhsiao committed:
wrapLine

Now uses binary search for the initial building of string for wrap check

hhsiao committed:
Update TBuffer.cpp

bug fix to handle line with multiple new line

hhsiao committed:
Removed redundant code

Removed redundant code

hhsiao committed:
Merge branch 'development' into Text-Wrapping

hhsiao commented:
Optimization has been done, ready for review

Kebap commented on src/ui/profile_preferences.ui:
```diff
@@ -1014,7 +1017,21 @@ you can use it but there could be issues with aligning columns of text</string>
                </property>
                <property name="text">
-                <string>Wrap lines at:</string>
+                <string>Wrap lines:</string>
+               </property>
+              </widget>
+             </item>
+             <item>
+              <widget class="QCheckBox" name="checkBox_autoWrap">
+               <property name="text">
+                <string>automatically</string>
+               </property>
+              </widget>
+             </item>
+             <item>
+              <widget class="QLabel" name="label_6">
+               <property name="text">
+                <string>or at</string>
```
Please find a way to combine these strings into one:
- Wrap lines
- automatically
- or at
- (number)

Because they will be translated into multiple languages that may or may not follow a same grammar

hhsiao committed:
Added containNewLine option to wrapLine

containNewLine indicate whether wrap line needs to check for newLine, this avoids 3 call to indexOf

hhsiao committed:
Merge branch 'Text-Wrapping' of https://github.com/hhsiao/Mudlet into Text-Wrapping

hhsiao committed:
Merge branch 'development' into Text-Wrapping

hhsiao committed:
Combined several block of codes

Combined several block of code for onlyWrapOneLine

hhsiao committed:
binarySearch function

two variant of binarySearch extracted into a function

SlySven commented on src/TBuffer.cpp:
```diff
@@ -2611,110 +2328,245 @@ void TBuffer::logRemainingOutput()
 
-// returns how many new lines have been inserted by the wrapping action
-int TBuffer::wrapLine(int startLine, int screenWidth, int indentSize, TChar& format)
+inline int TBuffer::binarySearchHorizontalAdvance(const int& lineIndex, const int& indentSize, const QString& lineIndent, const int& screenWidth, const int& subStringStart, const int& lineCharTotal, const QFontMetrics& qfm, int& lineCharIterator, const bool isBefore)
```
If this function/method has a return type of `int` then it should `return` something, as it doesn't this definition (and the declaration in the `TBuffer.h` file) had better be changed to:
```suggestion
inline void TBuffer::binarySearchHorizontalAdvance(const int& lineIndex, const int& indentSize, const QString& lineIndent, const int& screenWidth, const int& subStringStart, const int& lineCharTotal, const QFontMetrics& qfm, int& lineCharIterator, const bool isBefore)
```

hhsiao replied:
good catch!  Ha I meant to return an int but changed my mind.

hhsiao committed:
Change return value for binarySearchHorizontalAdvance

binarySearchHorizontalAdvance doesn't actually return anything, change it from int to void for return type.

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

hhsiao committed:
Corresponding changes to TBuffer.h

Previous suggested change left out the header file change.

hhsiao committed:
Simplified the wrapLine function

wrapLine used to go beyond the wrap point and then go back, it now cut it exactly where it's supposed to be

hhsiao committed:
Update TBuffer.cpp

Remove variables that isn't needed
Use QChar::LineFeed instead of "\n"

hhsiao committed:
Optimization for binarySearchHorizontal

Now check whether width can be calculated mathematically (using size * charWidth * factor) instead of drawing it with horizontalAdvance.  The best case scenario is horizontalAdvance is only called once.

hhsiao commented:
Done with optimization for this PR, ready for review

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2052,234 +2052,50 @@ void TBuffer::resetColors()
-    int last = buffer.size() - 1;
-    if (last < 0) {
+
+    // Prevent instance where buffer is empty
```
```suggestion
    // Prevent buffer from being empty
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2052,234 +2052,50 @@ void TBuffer::resetColors()
-    if (length < 1) {
-        return;
     }
-    length = std::min(length, MAX_CHARACTERS_PER_ECHO);
```
I'd keep this in. We've previously someone had try to echo a million characters with no line endings in Mudlet: https://discord.com/channels/283581582550237184/283582439002210305/689496621745963030

Haven't had any complaints since putting the limit in, so it seems to be doing its job.

hhsiao replied:
This should be fine, but I'll take a look.

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2052,234 +2052,50 @@ void TBuffer::resetColors()
-            firstChar = false;
-        }
+    
+    int counter = 0;
```
This does not seem to be used, right?

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2052,234 +2052,50 @@ void TBuffer::resetColors()
+    QStringRef lineText = text.midRef(sub_start, sub_end - sub_start);
+    lineBuffer.back().append(lineText);
+    const QFont hostFont = mpHost->getDisplayFont();
+    int mFontWidth = QFontMetrics(hostFont).averageCharWidth();
```
`m` stands for member variable of a class, so in this case I'd just go with -

```suggestion
    int fontWidth = QFontMetrics(hostFont).averageCharWidth();
```

Also could `fontWidth` and `wrapByPixel` also pass as `const`?

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2052,234 +2052,50 @@ void TBuffer::resetColors()
+    int mFontWidth = QFontMetrics(hostFont).averageCharWidth();
+    int wrapByPixel = mWrapAt * mFontWidth;
+    QFontMetrics qfm(hostFont);
+    if (qfm.horizontalAdvance(lineBuffer.back()) > wrapByPixel || lineText.indexOf("\n") != -1) {
```
```suggestion
    if (qfm.horizontalAdvance(lineBuffer.back()) > wrapByPixel || lineText.indexOf(QChar::LineFeed) > -1) {
```

Maybe this is faster/more intuitive?

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2460,107 +2276,8 @@ inline int TBuffer::skipSpacesAtBeginOfLine(const int row, const int column)
 
 inline int TBuffer::wrap(int startLine)
```
```suggestion
// wraps text all the way from startLine until the end of the buffer
inline int TBuffer::wrap(int startLine)
```

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2611,110 +2328,236 @@ void TBuffer::logRemainingOutput()
+}
+
+// The difference between this and wrap is, wrap does it from startLine all the way to the end
+// while wrapLine only does it for one line.
```
```suggestion
// wraps text for startLine only
```

vadi2 commented:
Awesome, reviewing it.

I'll also enable the clang-tidy review bot, it might have some suggestions. You've done so much work on this already, so I can apply those for you in case your battery is running low =)

vadi2 committed:
Merge branch 'development' into Text-Wrapping

vadi2 commented on src/TBuffer.cpp:
```diff
@@ -2613,110 +2330,236 @@ void TBuffer::logRemainingOutput()
+
+// The difference between this and wrap is, wrap does it from startLine all the way to the end
+// while wrapLine only does it for one line.
+// Thus, wrap uses pop_back and push_back while wrapLine uses remove_at and insert.
```
Is this comment still accurate?

hhsiao replied:
Actually, you are right, I'll change the comment

vadi2 commented on src/TBuffer.h:
```diff
@@ -147,7 +147,7 @@ class TBuffer
     bool insertInLine(QPoint& cursor, const QString& what, const TChar& format);
     void expandLine(int y, int count, TChar&);
-    int wrapLine(int startLine, int screenWidth, int indentSize, TChar& format);
+    int wrapLine(int startLine, int screenWidth, int indentSize, TChar& format, bool onlyWrapOneLine = true, bool containNewLine = false);
```
```suggestion
    int wrapLine(int startLine, int screenWidth, int indentSize, TChar& format, bool onlyWrapOneLine = true, bool containsNewLine = false);
```

vadi2 committed:
Reduce nesting in binarySearch

vadi2 committed:
QStrings are initialised to "" by default

vadi2 committed:
Remove unused variable

vadi2 committed:
Fix typo

vadi2 committed:
Comments

hhsiao committed:
Merge pull request #1 from vadi2/review-feedback

Minor review feedback

hhsiao committed:
More comment and variable rename

Added comment explaining what the variable does
Renamed variable to be more descriptive

hhsiao commented:
I've incorporated your recommendation (PR), and also addressed your questions.
Variable also renamed to be more descriptive.

hhsiao committed:
Fixed indention bug, comment fixing

Added suggested changes to the comment
Remove redundant variable
Fixed an indention bug

hhsiao committed:
Merge branch 'development' into Text-Wrapping

hhsiao committed:
Bug Fix

Fixed a bug where horizontal advance was being skipped due to truncation in the calculation

hhsiao committed:
Merge branch 'Text-Wrapping' of https://github.com/hhsiao/Mudlet into Text-Wrapping
